src/lib/group.ts

```diff
@@ -0,0 +1,394 @@
+
+  /**
+   * Coerces anything group-like to a {@link Group}.
+   */
+  constructor(g: {
+    x: FieldVar | Field | number | string | bigint;
+    y: FieldVar | Field | number | string | bigint;
+  });
+  /**
+   * Coerces anything group-like to a {@link Group}.
+   */
+  constructor(
+    x: FieldVar | Field | number | string | bigint,
+    y: FieldVar | Field | number | string | bigint
+  );
+  constructor(
+    arg1: GroupLike | FieldLike,
+    arg2: FieldLike | undefined = undefined
+  ) {
+    let x, y: FieldLike;
+
+    if (isGroupLike(arg1)) {
+      x = arg1.x;
+      y = arg1.y;
+    } else {
+      x = arg1 as FieldLike;
+      y = arg2 as FieldLike;
+    }
+
+    this.x = isField(x) ? x : new Field(x);
+    this.y = isField(y) ? y : new Field(y);
+
+    /*
+    technically elements are only group elements if they are on the curve (aka for two points y, x the following equation must hold y^2 = x^3 + 5 over our field)
+    but we never checked that in the original implementation
+
+
+    const { add, mul, sqrt, square } = Fp;
+
+    let onCurve = add(mul(xx, mul(xx, xx)), Pallas.b) === square(yy);
+    if (!onCurve) {
+      throw Error(`${{ xx, yy }} is not a valid group element`);
+    }
+    */
+  }
+
+  // helpers
+  static #toTuple(g: Group): [0, FieldVar, FieldVar] {
+    return [0, g.x.value, g.y.value];
+  }
+
+  static #fromAffine({ x, y }: { x: bigint; y: bigint; infinity: boolean }) {
+    return new Group({ x, y });
+  }
+
+  static #fromProjective({ x, y, z }: { x: bigint; y: bigint; z: bigint }) {
+    return this.#fromAffine(Pallas.toAffine({ x, y, z }));
+  }
+
+  #isConstant() {
+    return this.x.isConstant() && this.y.isConstant();
+  }
+
+  #toAffine() {
+    return Pallas.fromAffine({
+      x: this.x.toBigInt(),
+      y: this.y.toBigInt(),
+      infinity: false,
+    });
+  }
+
+  /**
+   * Return a {@link Bool} if the Group element is on the Pallas curve.
+   * It checks that the Weierstrass equation `y^2 = x^3 + 5` is satisfied.
+   */
+  onCurve() {
+    if (this.#isConstant()) {
+      const { add, mul, square } = Fp;
+      let y = this.y.toBigInt();
+      let x = this.x.toBigInt();
+      return Bool(add(mul(x, mul(x, x)), Pallas.b) === square(y));
+    } else {
+      let { x, y } = this;
+      // x^3 + 5 === y^2
+      let x3 = x.square().mul(x);
+      return x3.add(Pallas.b).equals(y.square());
+    }
+  }
+
+  /**
+   * Adds this {@link Group} element to another {@link Group} element.
+   *
+   * ```ts
+   * let g1 = Group({ x: 1, y: 1})
+   * let g2 = g1.add(g1)
+   * ```
+   */
+  add(g: Group) {
+    if (this.#isConstant() && g.#isConstant()) {
+      if (this.x.toBigInt() === 0n) {
+        return g;
+      } else if (g.x.toBigInt() === 0n) {
+        return this;
+      } else {
+        let g_proj = Pallas.add(this.#toAffine(), g.#toAffine());
+        return Group.#fromProjective(g_proj);
+      }
+    } else {
+      let [, x, y] = Snarky.group.add(Group.#toTuple(this), Group.#toTuple(g));
```

Comment:
I would like to move this entirely to JavaScript, but wanted to establish a baseline first